### ISSUE TITLE
Feature: Exit Conduit if pipeline config file fails to be provisioned (flag)

### ISSUE THREAD

**Issue by lovromazgon** [ConduitIO/conduit#850]
### Feature description

If a Conduit instance is deployed with the purpose to run only pipelines provisioned through pipeline config files, it would be nice if Conduit would stop right away if it detects that a pipeline config file is invalid (an invalid connector plugin config probably also makes the file invalid), otherwise it will seem like the server is healthy while some pipeline might actually not work.

We could add a flag that enables this functionality, e.g. `conduit -pipelines.exit-on-provision-error` (naming is not final of course).

**Comment by simonl2002** [comment]
Another option here is we add the flag as a `required` property on the pipeline in the config file. 

i.e.
```
version: 1.0
pipelines:
  file-to-file:
    status: running
    required: true
    ...
```

So conduit would exit if a pipeline with`required: true` fails to provision.

**Comment by hariso** [comment]
> 

@simonl2002 do you maybe have a specific use case in mind, that this would solve? I'm trying to think of a situation where we would want that some pipelines fail Conduit and others not.

**Comment by hariso** [comment]
@simonl2002 thank you for the reply!

My question was more directed to the solution you proposed, which is to add a flag on a pipeline in a config file (vs a global setting proposed initially in the issue).

That would mean that some pipelines would cause Conduit to exit if not provided, whereas others in the same file wouldn't.

**Comment by simonl2002** [comment]
@hariso that's correct. give granular control. having an command line arg removes any flexibility.

**Comment by lovromazgon** [comment]
What do you expect to happen if the file is invalid? For example, a required field is missing (or has a typo), a field name contains an invalid character, a field that we expect to contain a number contains a string etc. Currently, we ignore configuration files that are invalid and log an error, Conduit continues to run the remaining pipelines.

**Comment by simonl2002** [comment]
@lovromazgon hmmm I could see different arguments on what to do here. But I think maintaining current behavior makes the most sense. If pipeline file is invalid, ignore and log the error. I could see a world where we would want to make that behavior configurable but I don't really see the need to introduce that knob at the moment.

**Comment by lovromazgon** [comment]
Keeping the same behavior implies that Conduit will _not_ stop if the configuration file is invalid, because it won't read the configuration option `required: true`. In other words, the operator will still need to check the API if the pipeline is running and was provisioned with the right settings, and this feature is trying to ensure it doesn't have to do that, right?

**Comment by simonl2002** [comment]
@lovromazgon what happens if there are no valid pipeline files?

**Comment by lovromazgon** [comment]
Conduit simply starts without provisioning anything. It exposes an API so you can still interact with Conduit and create pipelines.

**Comment by simonl2002** [comment]
Conduit’s current operation is geared toward being a service that starts and presents an API. You can have pipelines defined and they may or may not successfully load, but that does not fundamentally affect the state of the conduit service.

Platform wants a different operational model in which Conduit primarily runs as a pipeline service. In this model if a pipeline doesn’t start/run then the service isn’t healthy/fully functional.

Questions:
* Do we want Conduit to exit if pipeline stops after being running for some period of time?
* Do we need to have an “Initializing” pipeline state? Currently there are two states “running” and “stopped.” Any init currently happens in running state.
* What does it mean for the pipeline to successful started. Open method can take seconds. At what point does a transition to stopped mean we should exit?
* What do we do in the N pipeline case if we’re not using a “required” flag?
* Some errors are only noticed upon records flowing through the pipeline. Would this be considered invalid configuration?

@lyuboxa and thoughts on these questions.

**Comment by lyuboxa** [comment]
> Some errors are only noticed upon records flowing through the pipeline. Would this be considered invalid configuration?

This question raises another, why keep conduit running when a pipeline fails? At this stage the server is considered degraded if N out of M pipelines have failed, the fix may be a reload with better config or restart the pipeline. Fiddling with the API is more work than restarting the process. 

So perhaps we should fail on error rather than provisioning.

**Comment by ahmeroxa** [comment]
IMHO I think the Conduit server should halt immediately if the configuration file has been provided and **any** part of it is invalid. I don't think we should introduce anything specific to enable this behavior (so no `required` field or additional command line flags).

My expectation is that I have provided a configuration file, so I expect the Conduit server to reflect what I have configured and **only** what I have configured. Running in a partial state would be surprising and may hide the problem (and possibly introduce other subtle issues that cause further confusion).
